lib/std/meta.zig

```diff
@@ -213,6 +213,124 @@ fn testSentinel() void {
+    testing.expect(ArrayChild(?[*]u8) == u8);
+    testing.expect(ArrayChild(?*[*]u8) == u8);
+    testing.expect(ArrayChild(?[]u8) == u8);
+    testing.expect(ArrayChild(?*[]u8) == u8);
+
+    testing.expect(ArrayChild(Vector(2, u8)) == u8);
+    testing.expect(ArrayChild(*Vector(2, u8)) == u8);
+    testing.expect(ArrayChild(?Vector(2, u8)) == u8);
+}
+
+/// Returns the given Slice or "Many Pointer" type with the same type except with the given sentinel value.
+pub fn Sentinel(comptime T: type, comptime sentinel_val: ArrayChild(T)) type {
+    return SentinelImpl(T, sentinel_val, false);
+}
+
+pub fn SentinelImpl(comptime T: type, comptime sentinel_val: ArrayChild(T), comptime in_child: bool) type {
+    const err = "unable to derive a sentinel pointer type from " ++ @typeName(T);
+    switch (@typeInfo(T)) {
+        .Array => |info| {
+            return @Type(builtin.TypeInfo { .Array = .{
+                .len = info.len,
+                .child = info.child,
+                .sentinel = sentinel_val,
+            }});
+        },
+        .Pointer => |info| {
+            switch (info.size) {
+                .One => {
+                    if (in_child) @compileError(err);
+                    return @Type(builtin.TypeInfo { .Pointer = .{
+                        .size = info.size,
+                        .is_const = info.is_const,
+                        .is_volatile = info.is_volatile,
+                        .alignment = info.alignment,
+                        .child = SentinelImpl(info.child, sentinel_val, true),
+                        .is_allowzero = info.is_allowzero,
+                        .sentinel = info.sentinel,
+                    }});
+                },
+                .C => @compileError("??? handle C pointers ???"),
+                .Many, .Slice => {
+                    return @Type(builtin.TypeInfo { .Pointer = .{
+                        .size = info.size,
+                        .is_const = info.is_const,
+                        .is_volatile = info.is_volatile,
+                        .alignment = info.alignment,
+                        .child = info.child,
+                        .is_allowzero = info.is_allowzero,
+                        .sentinel = sentinel_val,
+                    }});
+                },
+            }
+        },
+        else => @compileError(err),
+    }
+}
+
+/// Takes a Slice or Many Pointer and returns it with the Type modified to have the given sentinel value.
+/// This function assumes the caller has verified the memory contains the sentinel value.
+pub fn assumeSentinel(p: anytype, comptime sentinel_val: ArrayChild(@TypeOf(p))) Sentinel(@TypeOf(p), sentinel_val) {
+    const T = @TypeOf(p);
+    const err = "unable to derive a sentinel pointer type from " ++ @typeName(T);
+    const ReturnType = Sentinel(T, sentinel_val);
+    switch (@typeInfo(T)) {
+        .Pointer => |info| {
+            switch (info.size) {
+                .Slice => return @bitCast(ReturnType, p),
```

Comment:
Slices do not have a defined in-memory representation.

Reply:
I'm not sure how that's relevant here.  This is casting one slice type to another, the only difference being that the new type has a sentinel value.  This can't affect the in-memory representation as far as I know.